### PR TITLE
Proxy list

### PR DESCRIPTION
**Pull request by maZahaca**
* Added ability to set proxy and rotate it among the list
* Added description and typedef for PhantomEnvironment
* Dir ./parser_reports replaced with ./snapshots
* Added match and split transforms
* node version moved to 4.0.0

#58

### CODE REVIEW

**Comment by maZahaca** [comment]
@jifeon wtf with travis after update node to 4.0?
```bash
Uncaught Error: Cannot find module 'weak'
      at new D (node_modules/phantom/node_modules/dnode/index.js:28:20)
      at module.exports (node_modules/phantom/node_modules/dnode/index.js:8:12)
      at node_modules/phantom/phantom.js:159:13
      at Server.handler (node_modules/phantom/node_modules/shoe/index.js:22:9)
      at App.emit (node_modules/phantom/node_modules/shoe/node_modules/sockjs/lib/sockjs.js:182:27)
      at Session.emit_open (node_modules/phantom/node_modules/shoe/node_modules/sockjs/lib/transport.js:107:23)
```

File: lib/PhantomEnvironment.js [hunk 1]
```diff
@@ -313,6 +354,65 @@ PhantomEnvironment.prototype = _.create(Environment.prototype, /**@lends Phantom
...
+     * @returns {Promise}
+     * @private
+     */
+    _rotateProxy: function () {
+        var deferred = vow.defer();
+        var proxy = this._options.proxy;
+        var currentProxy = this._proxyCurrent;
+        if (proxy == undefined) {
+            deferred.resolve(null);
+            return deferred.promise();
```
Comment (jifeon) [inline lib/PhantomEnvironment.js]: `return vow.resolve(null)` and move `deferred` creation under this block

File: lib/PhantomEnvironment.js [hunk 2]
```diff
@@ -313,6 +354,65 @@ PhantomEnvironment.prototype = _.create(Environment.prototype, /**@lends Phantom
...
+        }
+        if (Array.isArray(proxy)) {
+            this._removeUnavailableProxy();
+            proxy = typeof this._options.proxyRotator === 'function' ?
+                this._options.proxyRotator.call(this, proxy, currentProxy) :
+                _.sample(proxy);
+
+            if (proxy == undefined) {
```
Comment (jifeon) [inline lib/PhantomEnvironment.js]: `!proxy`

**Comment by jifeon** [comment]
looks very cool!

**Comment by maZahaca** [comment]
@jifeon what can we do here about node version for travis? Downgrade?

**Pull request merged**